renter/contracts.go

```diff
@@ -0,0 +1,289 @@
+	}
+	hostAddr := settings.Address
+	vc, err := s.cm.TipContext()
+	if err != nil {
+		return rhp.Contract{}, nil, fmt.Errorf("failed to get validation context: %w", err)
+	}
+	startHeight := vc.Index.Height
+	if endHeight < startHeight {
+		return rhp.Contract{}, nil, errors.New("end height must be greater than start height")
+	}
+	renterAddr := s.wallet.Address()
+	if err != nil {
+		return rhp.Contract{}, nil, fmt.Errorf("failed to generate address: %w", err)
+	}
+	current := contract.Revision
+
+	// calculate the "base" storage cost to the renter and risked collateral for
+	// the host for the data already in the contract. If the contract height did
+	// not increase, base costs are zero.
+	var baseStorageCost, baseCollateral types.Currency
+	if contractEnd := endHeight + settings.WindowSize; contractEnd > contract.Revision.WindowEnd {
+		extension := contractEnd - contract.Revision.WindowEnd
+		baseStorageCost = settings.StoragePrice.Mul64(current.Filesize).Mul64(extension)
+		baseCollateral = settings.Collateral.Mul64(current.Filesize).Mul64(extension)
+	}
+
+	// calculate the total collateral the host is expected to add to the
+	// contract.
+	totalCollateral := baseCollateral.Add(additionalCollateral)
+	if totalCollateral.Cmp(settings.MaxCollateral) > 0 {
+		return rhp.Contract{}, nil, errors.New("collateral too large")
+	}
+
+	// create the renewed contract
+	//
+	// The host valid output includes the contract fee, base storage revenue,
+	// base collateral, and additional collateral. In the event of failure
+	// the base collateral and base storage cost should be burned.
+	validHostPayout := settings.ContractFee.Add(baseStorageCost).Add(totalCollateral)
+	missedHostPayout := settings.ContractFee.Add(additionalCollateral)
+	renewal := types.FileContract{
+		Filesize:           contract.Revision.Filesize,
+		FileMerkleRoot:     contract.Revision.FileMerkleRoot,
+		WindowStart:        endHeight,
+		WindowEnd:          endHeight + settings.WindowSize,
+		ValidRenterOutput:  types.SiacoinOutput{Address: renterAddr, Value: additionalRenterFunds},
+		ValidHostOutput:    types.SiacoinOutput{Address: hostAddr, Value: validHostPayout},
+		MissedRenterOutput: types.SiacoinOutput{Address: renterAddr, Value: additionalRenterFunds},
+		MissedHostOutput:   types.SiacoinOutput{Address: hostAddr, Value: missedHostPayout},
+		RenterPublicKey:    contract.Revision.RenterPublicKey,
+		HostPublicKey:      contract.Revision.HostPublicKey,
+	}
+
+	// create the clear and renew transaction. No signatures should be present
+	// yet.
+	renewalTxn := types.Transaction{
+		FileContractRevisions: []types.FileContractRevision{{
+			Parent:   contract.Parent,
+			Revision: rhp.ClearingRevision(contract.Revision),
+		}},
+		FileContracts: []types.FileContract{renewal},
+	}
+	// TODO: better fee calculation
+	renewalTxn.MinerFee = s.tpool.RecommendedFee().Mul64(vc.TransactionWeight(renewalTxn))
```

Comment:
RHP3 uses the transaction fees from the host's price table for renewal. If we use the renter's fee recommendation instead, should we also remove the fields from `HostSettings`?

Reply:
Maybe. I'm not sure what the intention was here. Blindly using whatever fee the host suggests is an obvious attack vector, so we would have to enforce a minimum and maximum fee anyway.